src/controllers/hid/hidcontroller.cpp

```diff
@@ -255,6 +255,35 @@ int HidController::close() {
     return 0;
 }
 
+QList<int> HidController::getInputReport(unsigned int reportID) {
+    Trace hidRead("HidController getInputReport");
+    unsigned char* pPreviousBuffer = m_pPollData[m_iPollingBufferIndex];
+    const int currentBufferIndex = (m_iPollingBufferIndex + 1) % kNumBuffers;
+    unsigned char* pCurrentBuffer = m_pPollData[currentBufferIndex];
+    int bytesRead;
+    pCurrentBuffer[0] = reportID;
+    bytesRead = hid_get_input_report(m_pHidDevice, pCurrentBuffer, kBufferSize);
+    controllerDebug(bytesRead << "bytes received by hid_get_input_report" << getName()
+                              << "serial #" << hid_serial
+                              << "(including report ID of" << reportID << ")");
+    bytesRead -= kReportIdSize;
+
+    if (bytesRead == m_iLastPollSize &&
```

Comment:
An empty `if` arm is unusual.

Moreover this code for buffering and de-duplicating the input should be extracted into a private member function instead of duplicating it. I suggest to return a `const char*` pointer to the incoming data on success and `nullptr` otherwise. Pass the actual hidapi call as a function pointer, the signatures of `hid_get_input_report` and `hid_read` should match.

Reply:
Or add an enum that you pass as argument and a switch case, might be easier to read.

Reply:
I moved buffering and de-duplicating in a private member function, as suggested.
Regarding merging getInputReport and getFeatureReport I decided against it, for multiple reasons:
- A function pointer would expose the symbols of the low level library below hidcontroler to one abstraction level above hidcontroller
- The argument list would differ between the C++ method and the JavaScript proxy method
- The internal buffer of de-duplicating code must never be used for a feature report. This would require copying of the buffer array.
- Code differences would require multiple branches inside the function